client-apis/encrypt.md

```diff
@@ -130,72 +180,180 @@ This document refers to the output of the key derivation algorithm as the derive
+this operation MUST calculate the encrypted content and an authentication tag using the
+[authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+specified by the [algorithm suite](../framework/algorithm-suites.md),
+with the following inputs:
+
+- The AAD is the serialized [message body AAD](../data-format/message-body-aad.md),
+  constructed as follows:
+  - The [message ID](../data-format/message-body-aad.md#message-id) is the same as the
+    [message ID](../data-frame/message-header.md#message-id) serialized in the header of this message.
+  - The [Body AAD Content](../data-format/message-body-aad.md#body-aad-content) depends on
+    whether the thing being encrypted is a regular frame or final frame.
+    Refer to [Message Body AAD](../data-format/message-body-aad.md) specification for more information.
+  - The [sequence number](../data-format/message-body-aad.md#sequence-number) is the sequence
+    number of the frame being encrypted.
+    If this is the first frame sequentially, this value MUST be 1.
+    Otherwise, this value MUST be 1 greater than the value of the sequence number
+    of the previous frame.
+  - The [content length](../data-format/message-body-aad.md#content-length) MUST have a value
+    equal to the length of the plaintext being encrypted.
+    - For a regular frame the length of this plaintext MUST equal the frame length.
+    - For a final frame this MUST be the length of the remaining plaintext bytes
+      which have not yet been encrypted,
+      whose length MUST be equal to or less than the frame length.
+- The IV is the [sequence number](../data-format/message-body-aad.md#sequence-number)
+  used in the message body AAD above,
+  padded to the [IV length](../data-format/message-header.md#iv-length).
+- The cipherkey is the derived data key
+- The plaintext is the next subsequence of consumable plaintext bytes that have not yet been encrypted.
+  - For a regular frame the length of this plaintext subsequence MUST equal the frame length.
+  - For a final frame this MUST be the remaining plaintext bytes which have not yet been encrypted,
+    whose length MUST be equal to or less than the frame length.
+
+This operation MUST serialize a regular frame or final frame with the following specifics:
+
+- [Sequence Number](../data-format/message-body.md#sequence-number): MUST be the sequence number of this frame,
+  as determined above.
+- [IV](../data-format/message-body.md#iv): MUST be the IV used when calculating the encrypted content above
+- [Encrypted Content](../data-format/message-body.md#encrypted-content): MUST be the encrypted content calculated above.
+- [Authentication Tag](../data-format/message-body.md#authentication-tag): MUST be the authentication tag
+  outputted when calculating the encrypted content above.
+
+Once the entire frame is serialized,
+the serialized frame MAY be released.
```

Comment:
I think the more important statement here is the implicit opposite: "The serialized frame MUST NOT be released before the entire frame is serialized". Perhaps it would be better to state that instead.

Given that, I'm not sure we need to weigh in on the MAY vs. SHOULD part on when it IS released.